File: problems/0209_minimum-size-subarray-sum/solution_prefix_bisect.py
from typing import List, Optional

from bisect import bisect_left


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int:
        n = len(nums)
        # prefix[i] = sum of the first i elements. Positivity makes it
        # strictly increasing, which is what licenses the binary search.
        prefix = [0] * (n + 1)
        for i, x in enumerate(nums):
            prefix[i + 1] = prefix[i] + x
        # Sentinel: an impossible length that survives when target is never met.
        best = n + 1
        for i in range(n):
            # The sum of nums[i..j) is prefix[j] - prefix[i], so the
            # shortest run starting at i reaching the target ends at the
            # first prefix >= prefix[i] + target.
            j = bisect_left(prefix, prefix[i] + target, i + 1, n + 1)
            if j <= n:
                best = min(best, j - i)
        return 0 if best == n + 1 else best

File: problems-bettercode/0201-0300/0209_minimum-size-subarray-sum/solution_prefix_bisect.cpp
class Solution {
  public:
    int minSubArrayLen(int target, vector<int> &nums) {
        int n = nums.size();
        // prefix[i] = sum of the first i elements. Positivity makes it
        // strictly increasing, which licenses the binary search; long longs
        // absorb prefix + target, which can pass 2^31.
        vector<long long> prefix(n + 1, 0);
        for (int i = 0; i < n; i++) {
            prefix[i + 1] = prefix[i] + nums[i];
        }
        // Sentinel: an impossible length that survives when target is never met.
        int best = n + 1;
        for (int i = 0; i < n; i++) {
            // Lower bound: the first prefix >= prefix[i] + target, searched
            // from i+1 on so the window has positive length.
            long long key = prefix[i] + target;
            int j = (int)(lower_bound(prefix.begin() + i + 1, prefix.end(), key) - prefix.begin());
            if (j <= n) {
                best = min(best, j - i);
            }
        }
        return best == n + 1 ? 0 : best;
    }
};
